source/error.c

```diff
@@ -0,0 +1,152 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/error.h>
+#include <aws/common/common.h>
+#include <assert.h>
+
+static AWS_THREAD_LOCAL int last_error = 0;
+
+static aws_error_handler global_handler = NULL;
+static void *global_error_context = NULL;
+
+static AWS_THREAD_LOCAL aws_error_handler thread_handler = NULL;
+AWS_THREAD_LOCAL void *thread_handler_context = NULL;
+
+#ifndef AWS_MAX_ERROR_SLOTS
+#define AWS_MAX_ERROR_SLOTS 16
+#endif
+
+/* to multiply by 16, you shift left by 5 */
+#define SLOT_MULTIPLIER_SHIFTS 0x05
```

Comment:
to multiply by 16, you shift left by 4

Reply:
ugh..... what I get for counting bits

Reply:
that was dumb anyways, since I moved it to a static const int, using normal multiplication is fine.